libs/pika/debugging/include/pika/debugging/demangle_helper.hpp

```diff
@@ -8,103 +8,76 @@
 #pragma once
 
 #include <pika/config.hpp>
-#include <cstdlib>
+#include <memory>
 #include <string>
-#include <type_traits>
 #include <typeinfo>
 
-// --------------------------------------------------------------------
-// Always present regardless of compiler : used by serialization code
+// gcc and clang both provide this header
+#if __has_include(<cxxabi.h>)
+# include <cxxabi.h>
+using cxxabi_supported__ = std::true_type;
+#else
+using cxxabi_supported__ = std::false_type;
+// create some dummy function to make the compiler happy in the true_type instantiation
+namespace abi {
+    template <typename... Ts>
+    char* __cxa_demangle(Ts... ts)
+    {
+        return nullptr;
+    }
+}    // namespace abi
+#endif
+
 // --------------------------------------------------------------------
 namespace pika::debug::detail {
-    template <typename T>
+    // default : use built-in typeid to get the best info we can
+    template <typename T, typename Enabled = std::false_type>
     struct demangle_helper
     {
         char const* type_id() const { return typeid(T).name(); }
     };
-}    // namespace pika::debug::detail
-
-#if defined(__GNUG__)
 
-# include <cxxabi.h>
-# include <memory>
-# include <stdlib.h>
-
-// --------------------------------------------------------------------
-// if available : demangle an arbitrary c++ type using gnu utility
-// --------------------------------------------------------------------
-namespace pika::debug::detail {
+    // if available : demangle an arbitrary c++ type using gnu utility
     template <typename T>
-    class cxxabi_demangle_helper
+    struct demangle_helper<T, std::true_type>
     {
-    public:
-        cxxabi_demangle_helper()
+        demangle_helper()
           : demangled_{abi::__cxa_demangle(typeid(T).name(), nullptr, nullptr, nullptr), std::free}
         {
         }
 
         char const* type_id() const { return demangled_ ? demangled_.get() : typeid(T).name(); }
 
     private:
-        std::unique_ptr<char, void (*)(void*)> demangled_;
+        std::unique_ptr<char, decltype(&std::free)> demangled_;
     };
-}    // namespace pika::debug::detail
-
-#else
 
-namespace pika::debug::detail {
     template <typename T>
-    using cxxabi_demangle_helper = demangle_helper<T>;
+    using cxx_type_id = demangle_helper<T, cxxabi_supported__>;
```

Comment:
Is there a good reason not to do something like this:
```
#if defined(PIKA_DEBUG_DETAIL_HAVE_CXXABI) // define this further up
    // if available : demangle an arbitrary c++ type using gnu utility
    template <typename T>
    struct cxx_type_id<T>
    {
        cxx_type_id()
          : demangled_{abi::__cxa_demangle(typeid(T).name(), nullptr, nullptr, nullptr), std::free}
        {
        }

        char const* type_id() const { return demangled_ ? demangled_.get() : typeid(T).name(); }

    private:
        std::unique_ptr<char, decltype(&std::free)> demangled_;
    };
#else
    // default : use built-in typeid to get the best info we can
    template <typename T>
    struct cxx_type_id
    {
        char const* type_id() const { return typeid(T).name(); }
    };
#endif
```

or even
```
    template <typename T>
    struct cxx_type_id<T>
    {
#if defined(PIKA_DEBUG_DETAIL_HAVE_CXXABI) // define this further up  
        // if available : demangle an arbitrary c++ type using gnu utility     
        cxx_type_id()
          : demangled_{abi::__cxa_demangle(typeid(T).name(), nullptr, nullptr, nullptr), std::free}
        {
        }

        char const* type_id() const { return demangled_ ? demangled_.get() : typeid(T).name(); }

    private:
        std::unique_ptr<char, decltype(&std::free)> demangled_;
#else
        // default : use built-in typeid to get the best info we can
        char const* type_id() const { return typeid(T).name(); }
#endif
    };
```
?

These would avoid potential issues with defining `__cxa_demangle`.